design/thin-edge-actors-design.md

```diff
@@ -553,23 +583,273 @@ impl From<Msg2> for Msg {
+  /// Return `None` when all the senders to this mailbox have been dropped and all the messages consumed.
+  pub async fn next(&mut self) -> Option<M> {
+    // ...
+  }
+  
+  /// Turn on/off logging of the messages consumed from this mailbox.
+  /// 
+  /// The messages are logged when returned by the `next()` method.
+  pub fn log_inputs(&mut self, on: bool) {
+    // ...
+  }
+}
 
-#### Size of the channel buffers
+pub enum RuntimeMessage {
+  RuntimeError(RuntimeError),
+  Timeout,
+  Shutdown,
+  LogInputs(bool),
+  LogOutputs(bool),
+  // ...
+}
+```
 
-### Behavior
+The message sender of an actor is implemented as an `Address<M>`, the address of its mailbox.
 
-### Instantiation
+```rust
+/// Create a new mailbox with its address
+pub fn new_mailbox<M>(buffer: usize) -> (Mailbox<M>, Address<M>) {
+  // ...
+}
 
-### Discovery
+impl<M : From<RuntimeMessage>> Address<M> {
+  /// Send a message to the mailbox with that address
+  /// 
+  /// Await the messages is actually in the mailbox.
+  /// Fail when the mailbox has been dropped.
+  pub async fn send_message(&mut self, message: M) -> Result<(), ChannelError> {
+    // ...
+  }
 
-### Runtime
+  /// Clone this address which can then be used from another actor
+  fn clone(&self) -> Self {
+    // ...
+  }
+}
+```
+
+However, actors can not directly use the addresses of their peers.
+If an actor sends messages of type `A` to an actor that processes messages of type `B: From<A>`,
+then we do not want the source actor to be aware of the `B` message type,
+and still less the unrelated message types consumed by the target as `C: Into<B>`.
+Indeed, that `B` message type encompasses all the message kinds supported by the target,
+and these, as the `C` type, can be completely unrelated to the source business domain that is described by the `A` type.
+
+Instead of an `Address<B>`, the source must be provided a `Recipient<A>`,
+that wraps the `Address<B>` and casts the `A` messages into `B` values under the hood.
+Without such an adapter, the source actor would have to depend on the `B` type. 
+
+![recipients](diagrams/recipients.svg)
+
+This adaptation of addresses into recipients is done using an intermediate trait: `Sender<M>`.
+The `Recipient<M>` is just a convenient way to manipulate boxes of `dyn` values.
+
+```rust
+/// A recipient for messages of type `M`
+pub type Recipient<M> = Box<dyn Sender<M>>;
+
+#[async_trait]
+pub trait Sender<M>: 'static + Send + Sync {
+    /// Send a message to the recipient,
+    /// returning an error if the recipient is no more expecting messages
+    async fn send(&mut self, message: M) -> Result<(), ChannelError>;
+
+    /// Clone this sender in order to send messages to the same recipient from another actor
+    fn recipient_clone(&self) -> Recipient<M>;
+}
+
+/// An `Address<M>` is a `Recipient<N>` provided `N` implements `Into<M>`
+impl<M: Message> Address<M> {
+  pub fn as_recipient<N: Message + Into<M>>(&self) -> Recipient<N> {
```

Comment:
It would have been better if such type conversions are done using an externally pluggable adaptors rather than expecting an `impl of Address<M>`. Else, it feels like every `Address`es needs to be aware of what all it can be converted into, which would lead to tight coupling between components.

Reply:
I agree that there is still room for improvement regarding these adapters. Even more, this is a key design point we need to polish.

However, note the loose type constraint: `<N: Message + Into<M>>`. Note also that this method returns a `Recipient` i.e. a `dyn` object abstracting the concrete address and ready to be used by an other actor unaware of the `M` message type.

```rust
impl C8YHandler {
    pub fn as_measurement_consumer(&self) -> Recipient<Measurement> {
         self.address.as_recipient()
    }
    pub fn as_event_consumer(&self) -> Recipient<Event> {
         self.address.as_recipient()
    }
    pub fn as_alarm_consumer(&self) -> Recipient<Alarm> {
         self.address.as_recipient()
    }
}
```